### ISSUE TITLE
Random() function giving an error with LUW DB2 

### ISSUE THREAD

**Issue by manojredward** [GoogleCloudPlatform/professional-services-data-validator#1239]
Hi Team, 
We are comparing the data between LUW DB2 and BQ and our validation are in YAML files. While executing the ROW validation, we gave use_random_rows = True and random_row_batch_size = '1000' and it fails to execute the random rows selection.

**Points for your reference**
1. We are using Cloud Run Job to execute the DVT
2. Our input validation files are in YAML format and stored in GCS bucket
3. Our connection json is also in GCS and we are overwriting PSO_DV_CONN_HOME environment variable and use it to execute cloud run job

**Error:**
08/22/2024 05:13:45 PM-ERROR: Error (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: Statement Execute Failed: [IBM][CLI Driver][DB2] SQL0440N No authorized routine named "RANDOM" of type "" having compatible arguments was found. SQLSTATE=42884 SQLCODE=-440

**Input YAML file:**
result_handler:
  project_id: <project_id>
  table_id: <results_table_id>
  type: BigQuery
source: db2
target: bq
validations:
- case_insensitive_match: false
  comparison_fields:
  - cast: null
    field_alias: ADDR_TYPE_ID_cell_level_validation
    source_column: addr_type_id
    target_column: ADDR_TYPE_ID
  - cast: null
    field_alias: LANG_CD_cell_level_validation
    source_column: lang_cd
    target_column: LANG_CD
  - cast: null
    field_alias: ADDR_TYPE_TEXT_cell_level_validation
    source_column: addr_type_text
    target_column: ADDR_TYPE_TEXT
  filter_status: null
  filters: []
  format: table
  labels: []
  primary_keys:
  - cast: null
    field_alias: ADDR_TYPE_ID_pk
    source_column: addr_type_id
    target_column: ADDR_TYPE_ID
  - cast: null
    field_alias: LANG_CD_pk
    source_column: lang_cd
    target_column: LANG_CD
  - cast: null
    field_alias: ADDR_TYPE_TEXT_pk
    source_column: addr_type_text
    target_column: ADDR_TYPE_TEXT
  **random_row_batch_size: '1000'**
  schema_name: DB2I0
  table_name: TPY_ADDR_TYPE_DESC
  target_schema_name: dp-dev-derived-ea49.product_ref_geographic
  target_table_name: address_type_desc
  threshold: 0
  trim_string_pks: false
  type: Row
  **use_random_rows: true**

**Comment by helensilva14** [comment]
Hi @manojredward! What is the version of your DB2 instance? The error is indicating it could not found a RANDOM function to use without parameters being given (https://www.ibm.com/docs/en/db2/9.7?topic=messages-sql0250-sql0499#sql0440n). 

I checked in our DVT test instance (LINUXX8664 11.5.8.0) and it works this way by using a default seed number, according to the documentation (https://www.ibm.com/docs/en/db2/11.5?topic=module-random-function).

![Screenshot 2024-08-23 at 17 15 32](https://github.com/user-attachments/assets/3ab459e9-d923-42c6-a7e5-c25ed175dbe6)

But as we searched from version 11.1 and before, the function requires a seed number as parameter:

- 11.0: https://www.ibm.com/docs/en/db2/11.1?topic=SSEPGG_11.1.0/com.ibm.dwe.cubemdx.doc/mdx_random.htm
- 10.5: https://www.ibm.com/docs/en/db2/10.5?topic=functions-random-function-mdx

If that's your case we indeed need to make a code change, thanks for reaching out!

**Comment by manojredward** [comment]
Thanks @helensilva14 for your response. 
I will check the DB2 version and get back to you. Meanwhile, Could you please guide me how to pass the seed value in YAML file. Right now, we are passing something like mentioned below,

random_row_batch_size: '1000'
type: Row
use_random_rows: true

**Comment by helensilva14** [comment]
As of now, it is not possible to pass a seed number either using DVT CLI or YAML config file.

We'll work on an initial version of a code change to pass a seed direct in our code and ask your help to test it soon!

**Comment by manojredward** [comment]
Hello @helensilva14 
We are using the below DB2 version, could you please check this?
**AIX v11.5.9.0
Mainframe V12R1M505**

**Comment by helensilva14** [comment]
Hi @manojredward! Can you please try to execute the following SQL queries in your instance and let us know if they run without errors? Below you can see the output in our DB2 instance (LINUXX8664 11.5.8.0):

- Without seed (using default value): `db2 "select random() from sysibm.sysdummy1;"`
- With seed value: `db2 "select random(2) from sysibm.sysdummy1;"`

![image](https://github.com/user-attachments/assets/1ca22e98-33ac-4406-9afc-14fd25be4721)

**Comment by manojredward** [comment]
Thanks @helensilva14, I will check and let you know asap. We are using AIX v11.5.9.0 DB2, hope this is fine. If we need to install any packages, please let us know.

**Comment by manojredward** [comment]
Hi @helensilva14,
We tried to run the below queries in our AIX DB2 and got an error as
**SQL Error [42884]: NO AUTHORIZED RANDOM NAMED  HAVING COMPATIBLE ARGUMENTS WAS FOUND. SQLCODE=-440, SQLSTATE=42884, DRIVER=4.33.31**

Attached image for your reference
<img width="666" alt="random_sysdummy" src="https://github.com/user-attachments/assets/f9744ac8-d232-4566-8adb-04d984a79716">
<img width="666" alt="random_2_sysdummy" src="https://github.com/user-attachments/assets/ce4c7fcf-8dcd-4b70-83c2-5e27beeaa796">

**Comment by helensilva14** [comment]
Hi @manojredward, thanks for sharing the error message and the images!

We did more research and might be the case that your instance/flavor only recognizes the function as `RAND()`, such as in this example: https://db2-sql-cookbook.org/#_getting_n_random_rows

Could you please try to execute the following query versions and let us know if they work?

- `select rand() from sysibm.sysdummy1;`
- `select rand(2) from sysibm.sysdummy1;`

**Comment by manojredward** [comment]
Hi @helensilva14 
We have tried with rand() command and it is working as expected in our DB2. Please find the screenshot below.
<img width="960" alt="image (10)" src="https://github.com/user-attachments/assets/4015f9bd-c44b-4baa-a19a-d12ca5cc8633">
<img width="960" alt="image (11)" src="https://github.com/user-attachments/assets/94b4ed21-8934-44e0-b74b-93bb5738a148">

**Comment by helensilva14** [comment]
Great to hear it! We updated our code to always use RAND() since this method is always available in all DB2 versions, which is not the case for RANDOM().

**Comment by manojredward** [comment]
Great, Thanks, Please let us know once the code is merged, we will recreate image and do our testing.

**Comment by helensilva14** [comment]
We merged the fix a few minutes ago, so it is available at the develop branch now. Can you try packaging the latest develop branch and testing? Please follow these steps:

1. Clone our DVT GitHub repo: `git clone https://github.com/GoogleCloudPlatform/professional-services-data-validator.git`
1. Enter in the folder and activate a Python virtual environment 
1. Package the develop the branch with `pip install . ` to package the current directory with the fix.
1. Test out the validation commands.
1. Install missing libraries, if needed

**Comment by manojredward** [comment]
@helensilva14 

By following the thread, we were actually cloned the code base from **branch: 1239-random-function-error-luw-db2** and created an image and executed our validation and we are getting the below error.

Error (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: Binding Error: [IBM][CLI Driver][DB2] SQL0440N  No authorized routine named "SETSEED" of type "" having compatible arguments was found.  SQLSTATE=42884 SQLCODE=-440

We have added the below changes to our yaml file,
**random_row_batch_size: '2'
use_random_rows: true**

Could you please take a look on **seed function** and let us know the updates.

**Comment by helensilva14** [comment]
Hi @manojredward, thank you very much for your testing! I didn't face this error in our instance.

Anyways, I added a new line to our code to always set a default seed value before using the RAND() function, you can check the new pull request here -> #1249 

Could you please repeat the steps and test again the branch 1239-random-function-error-luw-db2?

**Comment by manojredward** [comment]
Thanks @helensilva14 for making those changes, I will recreate the image with this branch and test.

**Comment by manojredward** [comment]
Hi @helensilva14, We have tested the code from `issue1239-random-function-error-luw-db2` branch with two yaml's and both are give me the same error.

**YAML #1:**
random_row_batch_size: '2'
type: Row
use_random_rows: true

**YAML #2:**
random_row_batch_size: null
type: Row
use_random_rows: true

**Error:**
09/03/2024 12:48:33 PM-ERROR: Error (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: Statement Execute Failed: [IBM][CLI Driver][DB2] SQL0583N  The use of routine or expression "RAND()" is invalid because it is not deterministic or has an external action.  SQLSTATE=42845 SQLCODE=-583

**Comment by helensilva14** [comment]
Hello @manojredward, thanks for reporting this new error! 

I tried a new approach of setting the seed value for the RAND() function itself and not externally.

Could you please repeat the steps and test again the branch `issue1239-random-function-error-luw-db2`?

**Comment by manojredward** [comment]
Hi @helensilva14, Please find the below error when we test with the updated code from `issue1239-random-function-error-luw-db2`.

**Error:**
09/05/2024 06:49:34 AM-ERROR: Error (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: Binding Error: [IBM][CLI Driver][DB2] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.  SQLSTATE=42610 SQLCODE=-418

**YAML:**
random_row_batch_size: '2'
type: Row
use_random_rows: true

**Comment by manojredward** [comment]
Hello @helensilva14 and @luispavaogoogle 
Just want to check, do you have any updates for me?

**Comment by helensilva14** [comment]
Hi @manojredward, sorry for taking so long to provide a response, I had a PTO in this meantime and I wasn't able to give you updates before that, apologies!

In our current pull request (#1249) I added a CAST to the seed value we are setting as default now and it might solve the latest error of `09/05/2024 06:49:34 AM-ERROR: Error (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: Binding Error: [IBM][CLI Driver][DB2] SQL0418N The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value. SQLSTATE=42610 SQLCODE=-418`. Link I took as a reference: https://www.zobristinc.com/blog/db2-sql-error-sqlcode-418-sqlstate-42610/

Could you please repeat the steps and test again the branch `issue1239-random-function-error-luw-db2`? Thanks!

**Comment by Rushikesh-Hande** [comment]
Hello @helensilva14,  I am working with @manojredward on testing this. We tried testing with the latest pull from the branch issue1239-random-function-error-luw-db2 but it failed with similar error

Error (ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: Binding Error: [IBM][CLI Driver][DB2] SQL0418N  The statement was not processed because the statement contains an invalid use of one of the following: an untyped parameter marker, the DEFAULT keyword, or a null value.  SQLSTATE=42610 SQLCODE=-418

If you could help to check on this?

**Comment by helensilva14** [comment]
Hi @Rushikesh-Hande! I was discussing with the other DVT members about your scenario here and how many attempts we already had, so we all agreed it would more efficient for both sides if we try to schedule a troubleshoot/debugging call. It will enable us to check things in real-time and get this issue solved as soon as possible, because our main goal is to unblock your usage!

Could any of you please send an email to data-validator-eng@google.com with your availability slots for this week still and the next one? Please don't forget to provide timezone information. Thank you!

**Comment by manojredward** [comment]
Hello @helensilva14, Please find the attached YAML file that we are using.
[row_level_validation_test.txt](https://github.com/user-attachments/files/17117300/row_level_validation_test.txt)

**Comment by helensilva14** [comment]
Thank you very much @manojredward! While talking to @luispavaogoogle, he asked if you ever tried to run the validations inverting the order, so instead of DB2 as source and BQ as target, you would set BQ as source and DB2 as target.

We are wondering if this might produce a different output, is it possible for you to test it?

**Comment by manojredward** [comment]
Sure @helensilva14, We can do that testing and let you know the output.

**Comment by sundar-mudupalli-work** [comment]
Hi,

@helensilva14 - that is a great idea - at least the random function step will work in BigQuery.

If Helen's suggestion does not solve your problem, you can try to figure out how to get the query to find the random rows to execute in DB2. The statement the DVT uses is given below:
```
SELECT t0.addr_type_text, t0.lang_cd FROM (SELECT t1.addr_type_text AS addr_type_text, t1.lang_cd AS lang_cd FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 ORDER BY rand() ASC LIMIT 25
```
You can check to see if this query runs in DBeaver. It should produce 25 pairs of primary keys, for `addr_type_text` and `lang_cd`. If it does not succeed, then something to troubleshoot and fix. If it does work in DBeaver, check if DVT can run this query as shown below: 
```
data-validation query -c db2 -q 'SELECT t0.addr_type_text, t0.lang_cd FROM (SELECT t1.addr_type_text AS addr_type_text, t1.lang_cd AS lang_cd FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 ORDER BY rand() ASC LIMIT 25'
```
If it fails, you will need to figure out why it works with DBeaver, but not with DVT. Maybe the user does not have privileges to run RAND() or RAND() is not available.

Thanks.

Sundar Mudupalli

**Comment by manojredward** [comment]
Hi @helensilva14 
I have checked with swapping source and target and now i able to execute the row level validation. However i am unable to restrict the number of rows to test.

I have tested with a particular table and total number of rows in that table is 22 and execute DVT with random_row_batch_size = '2' and  random_row_batch_size = '1' and both are generated 66 rows in the target BQ table.

It looks like eventhough I have passed random_row_batch_size, it is executing all the records from the table.

Please find the attached YAML test file that I have used to test.
[src_bq_target_db2.txt](https://github.com/user-attachments/files/17123297/src_bq_target_db2.txt)

**Comment by manojredward** [comment]
Hi @sundar-mudupalli-work 

I have tried to execute the below query and it failed in DBeaver.

`SELECT t0.addr_type_text, t0.lang_cd FROM (SELECT t1.addr_type_text AS addr_type_text, t1.lang_cd AS lang_cd FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 ORDER BY rand() ASC LIMIT 25`

**Error:**
`SQL Error [42845]: THE USE OF FUNCTION OR EXPRESSION RAND() IS INVALID BECAUSE IT IS NOT DETERMINISTIC OR HAS AN EXTERNAL ACTION. SQLCODE=-583, SQLSTATE=42845, DRIVER=4.33.31`

**Comment by sundar-mudupalli-work** [comment]
> data-validation query -c db2 -q 'SELECT t0.addr_type_text, t0.lang_cd FROM (SELECT t1.addr_type_text AS addr_type_text, t1.lang_cd AS lang_cd FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 ORDER BY rand() ASC LIMIT 25'

Since I don't have access to db2 on z/os - can you execute this query on DBeaver (DB2) to see if it works:
```
data-validation query -c db2 -q 'SELECT t0.addr_type_text, t0.lang_cd FROM (SELECT t1.addr_type_text AS addr_type_text, t1.lang_cd AS lang_cd, RAND() as randnum FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 ORDER BY randnum ASC LIMIT 25'
```
Thank you.

**Comment by sundar-mudupalli-work** [comment]
> Hi @helensilva14 I have checked with swapping source and target and now i able to execute the row level validation. However i am unable to restrict the number of rows to test.
> 
> I have tested with a particular table and total number of rows in that table is 22 and execute DVT with random_row_batch_size = '2' and random_row_batch_size = '1' and both are generated 66 rows in the target BQ table.
> 
> It looks like eventhough I have passed random_row_batch_size, it is executing all the records from the table.
> 
> Please find the attached YAML test file that I have used to test. [src_bq_target_db2.txt](https://github.com/user-attachments/files/17123297/src_bq_target_db2.txt)

This is really odd - I am able to get the same thing to work in my environment. Can you try self row validation in BigQuery to see if the problem can be reproduced without DB2. For example (please check for typos)
```
data-validation validate row -sc bq -tc bq -tbls=dp-dev-derived-ea49.product_ref_insur_products.auto_veh_cnt -pk=LANG_CD,VEH_CNT_GRP,GRP_MIN_VALUE,GRP_MAX_VALUE,VEH_CNT_SHRTDESC,VEH_CNT_DESC -comp-fields=LANG_CD,VEH_CNT_GRP,GRP_MIN_VALUE,GRP_MAX_VALUE,VEH_CNT_SHRTDESC,VEH_CNT_DESC -rr -rbs 2
```
Thank you.

**Comment by manojredward** [comment]
> > Hi @helensilva14 I have checked with swapping source and target and now i able to execute the row level validation. However i am unable to restrict the number of rows to test.
> > I have tested with a particular table and total number of rows in that table is 22 and execute DVT with random_row_batch_size = '2' and random_row_batch_size = '1' and both are generated 66 rows in the target BQ table.
> > It looks like eventhough I have passed random_row_batch_size, it is executing all the records from the table.
> > Please find the attached YAML test file that I have used to test. [src_bq_target_db2.txt](https://github.com/user-attachments/files/17123297/src_bq_target_db2.txt)
> 
> This is really odd - I am able to get the same thing to work in my environment. Can you try self row validation in BigQuery to see if the problem can be reproduced without DB2. For example (please check for typos)
> 
> ```
> data-validation validate row -sc bq -tc bq -tbls=dp-dev-derived-ea49.product_ref_insur_products.auto_veh_cnt -pk=LANG_CD,VEH_CNT_GRP,GRP_MIN_VALUE,GRP_MAX_VALUE,VEH_CNT_SHRTDESC,VEH_CNT_DESC -comp-fields=LANG_CD,VEH_CNT_GRP,GRP_MIN_VALUE,GRP_MAX_VALUE,VEH_CNT_SHRTDESC,VEH_CNT_DESC -rr -rbs 2
> ```
> 
> Thank you.

Hi @sundar-mudupalli-work 
I have tried to execute the row validation directly with the below command and getting an error.
**Original command:**
`data-validation validate row -sc bq -tc bq -tbls=dp-dev-derived-ea49.product_ref_insur_products.auto_veh_cnt -pk=LANG_CD,VEH_CNT_GRP,GRP_MIN_VALUE,GRP_MAX_VALUE,VEH_CNT_SHRTDESC,VEH_CNT_DESC -comp-fields=LANG_CD,VEH_CNT_GRP,GRP_MIN_VALUE,GRP_MAX_VALUE,VEH_CNT_SHRTDESC,VEH_CNT_DESC -rr -rbs 2`
**Error:**
`ibis.common.exceptions.IntegrityError: Duplicate column name(s): ['LANG_CD', 'VEH_CNT_GRP', 'GRP_MIN_VALUE', 'GRP_MAX_VALUE', 'VEH_CNT_SHRTDESC', 'VEH_CNT_DESC']`

Them, I have changed the data-validation command and able to execute that. However, it is not executing the random rows instead it is taking all the rows and doing the random validation.
**Updated command:**
`data-validation validate row -sc bq -tc bq -tbls=dp-dev-derived-ea49.product_ref_insur_products.auto_veh_cnt -pk=LANG_CD,VEH_CNT_GRP,GRP_MIN_VALUE -comp-fields=GRP_MAX_VALUE,VEH_CNT_SHRTDESC,VEH_CNT_DESC -rr -rbs 2`

**Comment by helensilva14** [comment]
Hi @manojredward, please try to execute the following YAML:

```yaml
result_handler:
  project_id: dp-dev-processing-4993
  table_id: cdw_dvt_validation.dvt_results_all_tables_sample
  type: BigQuery
source: bq
target: bq
validations:
- case_insensitive_match: false
  comparison_fields:
  - cast: null
    field_alias: LANG_CD_cell_level_validation
    source_column: LANG_CD
    target_column: LANG_CD
  - cast: null
    field_alias: VEH_CNT_GRP_cell_level_validation
    source_column: VEH_CNT_GRP
    target_column: VEH_CNT_GRP
  - cast: null
    field_alias: GRP_MIN_VALUE_cell_level_validation
    source_column: GRP_MIN_VALUE
    target_column: GRP_MIN_VALUE
  - cast: null
    field_alias: GRP_MAX_VALUE_cell_level_validation
    source_column: GRP_MAX_VALUE
    target_column: GRP_MAX_VALUE
  - cast: null
    field_alias: VEH_CNT_SHRTDESC_cell_level_validation
    source_column: VEH_CNT_SHRTDESC
    target_column: VEH_CNT_SHRTDESC
  - cast: null
    field_alias: VEH_CNT_DESC_cell_level_validation
    source_column: VEH_CNT_DESC
    target_column: VEH_CNT_DESC
  filter_status: null
  filters: []
  format: table
  labels:
  - !!python/tuple
    - validation_type
    - cell_level_validation
  - !!python/tuple
    - dag_name
    - dvt_auto_reference_42_tables
  - !!python/tuple
    - stm_file
    - stm_auto_reference_42_tables.json
  primary_keys:
  - cast: null
    field_alias: LANG_CD_pk
    source_column: LANG_CD
    target_column: LANG_CD
  - cast: null
    field_alias: VEH_CNT_GRP_pk
    source_column: VEH_CNT_GRP
    target_column: VEH_CNT_GRP
  - cast: null
    field_alias: GRP_MIN_VALUE_pk
    source_column: GRP_MIN_VALUE
    target_column: GRP_MIN_VALUE
  - cast: null
    field_alias: GRP_MAX_VALUE_pk
    source_column: GRP_MAX_VALUE
    target_column: GRP_MAX_VALUE
  - cast: null
    field_alias: VEH_CNT_SHRTDESC_pk
    source_column: VEH_CNT_SHRTDESC
    target_column: VEH_CNT_SHRTDESC
  - cast: null
    field_alias: VEH_CNT_DESC_pk
    source_column: VEH_CNT_DESC
    target_column: VEH_CNT_DESC
  random_row_batch_size: '2'
  schema_name: dp-dev-derived-ea49.product_ref_insur_products
  table_name: auto_veh_cnt
  target_schema_name: dp-dev-derived-ea49.product_ref_insur_products
  target_table_name: auto_veh_cnt
  threshold: 0
  trim_string_pks: false
  type: Row
  use_random_rows: true
```

It is based on this other testing YAML below that I created to validate BQ to BQ random row execution. I was able to get the exact number of random rows I provided. 

```yaml
result_handler: {}
source: bq
target: bq
validations:
- case_insensitive_match: false
  comparison_fields:
  - cast: null
    field_alias: station_id_CELL_LEVEL_VALIDATION
    source_column: station_id
    target_column: station_id
  filter_status: null
  filters: []
  format: table
  labels: []
  primary_keys:
  - cast: null
    field_alias: station_id_PK
    source_column: station_id
    target_column: station_id
  random_row_batch_size: '2'
  schema_name: bigquery-public-data.new_york_citibike
  table_name: citibike_stations
  target_schema_name: bigquery-public-data.new_york_citibike
  target_table_name: citibike_stations
  threshold: 0.0
  type: Row
  use_random_rows: true
```

**Comment by manojredward** [comment]
Hi @helensilva14, Please find the details below,

BQ Table: auto_veh_cnt
DB2 Table: TPV_VEH_CNT
Total no. of rows in both table: 22

**Code from this Branch:**
https://github.com/GoogleCloudPlatform/professional-services-data-validator/tree/issue1239-random-function-error-luw-db2

Source: BQ
Target: BQ
src_bq_target_bq.yaml

**Scenario #1:**
random_row_batch_size = 5
132 new rows added

**Scenario #2:**
random_row_batch_size = 2
66 new rows added

**Code from this Master Branch:**
Source: BQ
Target: BQ
src_bq_target_bq.yaml

**Scenario #1:**
random_row_batch_size = 5
132 new rows added

**Scenario #2:**
random_row_batch_size = 2
66 new rows added

**YAML File:**
```
result_handler:
  project_id: dp-dev-processing-4993
  table_id: cdw_dvt_validation.dvt_results_row_v_issue
  type: BigQuery
source: bq
target: bq
validations:
- case_insensitive_match: false
  comparison_fields:
  - cast: null
    field_alias: LANG_CD_cell_level_validation
    source_column: LANG_CD
    target_column: LANG_CD
  - cast: null
    field_alias: VEH_CNT_GRP_cell_level_validation
    source_column: VEH_CNT_GRP
    target_column: VEH_CNT_GRP
  - cast: null
    field_alias: GRP_MIN_VALUE_cell_level_validation
    source_column: GRP_MIN_VALUE
    target_column: GRP_MIN_VALUE
  - cast: null
    field_alias: GRP_MAX_VALUE_cell_level_validation
    source_column: GRP_MAX_VALUE
    target_column: GRP_MAX_VALUE
  - cast: null
    field_alias: VEH_CNT_SHRTDESC_cell_level_validation
    source_column: VEH_CNT_SHRTDESC
    target_column: VEH_CNT_SHRTDESC
  - cast: null
    field_alias: VEH_CNT_DESC_cell_level_validation
    source_column: VEH_CNT_DESC
    target_column: VEH_CNT_DESC
  filter_status: null
  filters: []
  format: table
  labels:
  - !!python/tuple
    - validation_type
    - cell_level_validation
  - !!python/tuple
    - dag_name
    - dvt_auto_reference_42_tables
  - !!python/tuple
    - stm_file
    - stm_auto_reference_42_tables.json
  primary_keys:
  - cast: null
    field_alias: LANG_CD_pk
    source_column: LANG_CD
    target_column: LANG_CD
  - cast: null
    field_alias: VEH_CNT_GRP_pk
    source_column: VEH_CNT_GRP
    target_column: VEH_CNT_GRP
  - cast: null
    field_alias: GRP_MIN_VALUE_pk
    source_column: GRP_MIN_VALUE
    target_column: GRP_MIN_VALUE
  - cast: null
    field_alias: GRP_MAX_VALUE_pk
    source_column: GRP_MAX_VALUE
    target_column: GRP_MAX_VALUE
  - cast: null
    field_alias: VEH_CNT_SHRTDESC_pk
    source_column: VEH_CNT_SHRTDESC
    target_column: VEH_CNT_SHRTDESC
  - cast: null
    field_alias: VEH_CNT_DESC_pk
    source_column: VEH_CNT_DESC
    target_column: VEH_CNT_DESC
  random_row_batch_size: '5'
  schema_name: dp-dev-derived-ea49.product_ref_insur_products
  table_name: auto_veh_cnt
  target_schema_name: dp-dev-derived-ea49.product_ref_insur_products
  target_table_name: auto_veh_cnt
  threshold: 0
  trim_string_pks: false
  type: Row
  use_random_rows: true
```

**Comment by helensilva14** [comment]
Hi @manojredward, after chatting again with the team, we finally noticed the real root cause in here! 🚀 

<ins>**Even though you are informing 5 PKs, random rows is only being applied to the first PK (LANG_CD_pk)**</ins>, it's a known limitation we have and there is an open issue to be able to support multiple PKs, because it involves a lot more complexity to create the filters, please take a look here #774. 

So it seems your key LANG_CD_pk is not unique enough so that's why you are always getting more values than requested in `random_row_batch_size`. Being aware of it now, try to evaluate if this can be a problem for larger tables or if you could select other PK that could result in less rows.

Thanks again for your collaboration with us and apologies for not realizing it earlier.

**Comment by Rushikesh-Hande** [comment]
Thanks @helensilva14  for checking on this. So the limitation we've encountered concerning the inability to specify a combination of columns as a primary key is specific to BigQuery, or if it applies universally across all source and target systems(even for DB2 as a source) ?

**Comment by helensilva14** [comment]
It applies across all of our supported engines

**Comment by Rushikesh-Hande** [comment]
Thanks @helensilva14 for checking on this, whats a recommended way to do cell by cell row validation on sample records which is not having one columns as primary key ?

**Comment by helensilva14** [comment]
Hi @Rushikesh-Hande! You can still use this same approach of passing a PK and a random_row_batch_size, but first you need to make an analysis on which column would be the ideal PK in the sense of having more distinct values available.

Take a look of my example if I wanted to use this column `content` AS PK:

`db2 "SELECT content, COUNT(*) AS count FROM entries GROUP BY content;"`

```
CONTENT                       COUNT      
----------------------------------------
I got here!                    2
Me too!                        3
More data Coming               1
zippp!                         1
```

DVT is executing this command: `db2 "SELECT content FROM entries ORDER BY RAND() LIMIT 2;"`

```
CONTENT                                                                                           
----------------------
Me too!                                                    
Me too!
```

**Comment by sundar-mudupalli-work** [comment]
Hi,

Clarification - the original issue with not being able to use DVT Random Row in DB2 on z/OS is [due to the following limitation:](https://www.ibm.com/docs/en/db2-for-zos/12?topic=subselect-order-by-clause#db2z_sql_orderbyclause__rorder1)
```
The sort-key-expression cannot include an expression that is not deterministic or a function that is defined to have an external action except for the RID built-in function and the ROW CHANGE expression.
```
rand() used by DVT in this SQL expression to find the random primary keys is non-deterministic:
```
SELECT t0.addr_type_text FROM (SELECT t1.addr_type_text AS addr_type_text FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 ORDER BY rand() ASC LIMIT 25
```
This is not supported by DB2 on z/OS

Possible workaround - check if this SQL query is supported by DB2:
```
SELECT t0.addr_type_text FROM (SELECT t1.addr_type_text AS addr_type_text, RAND() as randnum FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 ORDER BY randnum ASC LIMIT 25'
```
If this works - we can change the DVT to use above SQL query instead of the earlier SQL query having the rand() function in order by clause.

Sundar Mudupalli

**Comment by manojredward** [comment]
Hello @sundar-mudupalli-work 
Thanks for checking this. Please find the error details after executing the queries that you specified above.

**Query1:**
```
SELECT t0.addr_type_text FROM 
(SELECT t1.addr_type_text AS addr_type_text FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 
ORDER BY rand() ASC LIMIT 25;
```
**Error:**
```
SQL Error [42845]: THE USE OF FUNCTION OR EXPRESSION RAND() IS INVALID
BECAUSE IT IS NOT DETERMINISTIC OR HAS AN EXTERNAL ACTION. 
SQLCODE=-583, SQLSTATE=42845, DRIVER=4.33.31
```
------------------------------------------------------------
**Query2:**
```
SELECT t0.addr_type_text FROM 
(SELECT t1.addr_type_text AS addr_type_text, 
RAND() as randnum FROM DB2I0."TPY_ADDR_TYPE_DESC" AS t1) AS t0 
ORDER BY randnum ASC LIMIT 25';
```
**Error:**
```
SQL Error [42603]: THE STRING CONSTANT BEGINNING  IS NOT TERMINATED. 
SQLCODE=-10, SQLSTATE=42603, DRIVER=4.33.31
```

**Comment by helensilva14** [comment]
> Great to hear it! We updated our code to always use RAND() since this method is always available in all DB2 versions, which is not the case for RANDOM().

Closing issue as completed because the initial problem was solved as quoted here. Everything else after that were other issues specific only to their on-prem database and it was not possible to replicate any of them.
